drivers/video/video_renesas_ra_ceu.c

```diff
@@ -0,0 +1,540 @@
+/*
+ * Copyright (c) 2025 Renesas Electronics Co.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT renesas_ra_ceu
+
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/video.h>
+#include <zephyr/irq.h>
+#include <zephyr/logging/log.h>
+#include <soc.h>
+#include <errno.h>
+
+#include "video_device.h"
+#include "r_ceu.h"
+
+LOG_MODULE_REGISTER(renesas_ra_video_ceu, CONFIG_VIDEO_LOG_LEVEL);
+
+/*
+ * Hardware alignment constraints:
+ * - Width: 128–2560 pixels, must be a multiple of 8
+ * - Height: 96–1920 lines, must be a multiple of 4
+ */
+#define VIDEO_RA_CEU_WIDTH_MIN    128U
+#define VIDEO_RA_CEU_WIDTH_MAX    2560U
+#define VIDEO_RA_CEU_WIDTH_ALIGN  8U
+#define VIDEO_RA_CEU_HEIGHT_MIN   96U
+#define VIDEO_RA_CEU_HEIGHT_MAX   1920U
+#define VIDEO_RA_CEU_HEIGHT_ALIGN 4U
+
+/*
+ * Default capture configuration:
+ * - Resolution: 128x96
+ * - Bytes per pixel: 2
+ */
+#define VIDEO_RA_CEU_DEFAULT_WIDTH           VIDEO_RA_CEU_WIDTH_MIN
+#define VIDEO_RA_CEU_DEFAULT_HEIGHT          VIDEO_RA_CEU_HEIGHT_MIN
+#define VIDEO_RA_CEU_DEFAULT_START_X         0U
+#define VIDEO_RA_CEU_DEFAULT_START_Y         0U
+#define VIDEO_RA_CEU_DEFAULT_BYTES_PER_PIXEL 2U
+
+struct video_renesas_ra_ceu_config {
+	void (*irq_config_func)(const struct device *dev);
+	const struct device *clock_dev;
+	const struct device *cam_xclk_dev;
+	const struct device *source_dev;
+	const struct pinctrl_dev_config *pincfg;
+	const struct clock_control_ra_subsys_cfg clock_subsys;
+};
+
+struct video_renesas_ra_ceu_data {
+	struct st_ceu_instance_ctrl *fsp_ctrl;
+	struct st_capture_cfg *fsp_cfg;
+	struct st_ceu_extended_cfg *fsp_extend_cfg;
+	struct video_format fmt;
+	struct k_fifo fifo_in;
+	struct k_fifo fifo_out;
+	atomic_ptr_t vbuf;
+	atomic_t streaming;
+#ifdef CONFIG_POLL
+	struct k_poll_signal *signal;
+#endif
+};
+
+extern void ceu_isr(void);
+
+static void video_renesas_ra_ceu_callback(capture_callback_args_t *p_args)
+{
+	const struct device *dev = p_args->p_context;
+	struct video_renesas_ra_ceu_data *data = dev->data;
+	struct video_buffer *curr_vbuf;
+	struct video_buffer *next_vbuf;
+	fsp_err_t err;
+
+	if (p_args->event & CEU_EVENT_FRAME_END) {
+		curr_vbuf = atomic_ptr_get(&data->vbuf);
+		if (curr_vbuf && atomic_ptr_cas(&data->vbuf, curr_vbuf, NULL)) {
+			curr_vbuf->timestamp = k_uptime_get_32();
+			k_fifo_put(&data->fifo_out, curr_vbuf);
+		}
```

Comment:
If wanting a complete support for the polling API, you may want to raise a completion signal here, or remove support for the polling API altogether (given this will be automated soon).

https://github.com/zephyrproject-rtos/zephyr/blob/5525941d839099309a44a0be868151fa894a5425/drivers/video/video_sw_generator.c#L312-L314